core/CMakeLists.txt

```diff
@@ -333,9 +333,9 @@ endif (WIN32)
 _DR_get_static_libc_list(static_libc_list)
 if (WIN32)
   if (DEBUG)
-    set(WIN32_C_LIB libcmtd)
+    set(WIN32_C_LIB libcmtd Psapi)
```

Comment:
Please go through all my review comments on the original PR #3328 and address them (or explain in a review comment why they do not apply) and when done re-request a review.  E.g. see https://github.com/DynamoRIO/dynamorio/pull/3328/files#r246456892 which seems to have been ignored in this PR.

Reply:
Will do. I made the mistake of assuming all resolved conversations are resolved...

For this specific comment, psapi is actually not used anymore, so I just removed it.
Addressed most of the other comments as well.
Would love some pointers on how to address the documentation and testing parts.

Reply:
> I made the mistake of assuming all resolved conversations are resolved

Hmm, did not realize some of those are resolved: that is strange.  Not sure if the original author recently resolved and then failed to push the changes or sthg.

> Would love some pointers on how to address the documentation and testing parts.

Starting by bringing the content from the prior comments on docs: from https://github.com/DynamoRIO/dynamorio/pull/3328#discussion_r246552961 here:

+ Maybe the main documentation ("page_deploy" in intro.dox) should talk about attaching as a supported feature?
+ It should be listed in the changelog in release.dox for sure as a major feature.

And on testing https://github.com/DynamoRIO/dynamorio/pull/3328#discussion_r246553938

+ Would it be possible to write some simple tests of these new attach features? It will take a little more work than adding a regular test because you'll need to launch something in the background. See client.nudge_test and how it has a .runall file and uses run_in_bg to run the infloop app in the background.
+ If you'd prefer to make adding a test a separate CL that would be fine.

Reply:
>* Maybe the main documentation ("page_deploy" in intro.dox) should talk about attaching as a supported feature?
>* It should be listed in the changelog in release.dox for sure as a major feature.

Added. Is it in the correct place?

>* Would it be possible to write some simple tests of these new attach features? It will take a little more work than adding a regular test because you'll need to launch something in the background. See client.nudge_test and how it has a .runall file and uses run_in_bg to run the infloop app in the background.
>* If you'd prefer to make adding a test a separate CL that would be fine.

Can you explain the structure of the nudge_test files? How does it use run_in_bg?

Reply:
> Can you explain the structure of the nudge_test files? How does it use run_in_bg?

See https://github.com/DynamoRIO/dynamorio/blob/master/suite/tests/CMakeLists.txt#L1304 and https://github.com/DynamoRIO/dynamorio/blob/master/suite/tests/runall.cmake#L32

Reply:
> Added. Is it in the correct place?

Yes.

Reply:
I'm getting lost with the tests (and for some reason I am unable to run them locally).

If I understand correctly, I need to add 3 files as in `nudge_test`, and then add to `suite\tests\CMakeLists.txt` something like:
`  tobuild_ci(client.attach_test client-interface/attach_test.runall "" "-attach <pid>" "")`
but I don't think I have the process id at that point...

What am I missing? Should I add a new type of runall?

Reply:
> but I don't think I have the process id at that point...

The nudge test also needs the pid and it gets it from a file that run_in_bg writes for it: look for "pidfile".and 

> for some reason I am unable to run them locally

Run "ctest -V -R nudge_test" in a build directory where BUILD_TESTS and DEBUG where both configured as ON.

Reply:
I've added the test. I think it creates the target process with `run_in_bg` properly (not under DR so we can test the attach), and runs an ok `drrun` attach command.
But for some reason it is still failing. Can you take a look?

Reply:
You're saying it passes locally but not on GA CI?  Supplying more detals on the failure and pasting the relevant logs would make it easier to get help -- I went and got them myself:

```
2021-09-10T23:43:19.4669625Z         Start 108: code_api|client.attach_test
2021-09-10T23:43:19.4670098Z 
2021-09-10T23:43:19.4675119Z 108: Test command: "C:\Program Files\CMake\bin\cmake.exe" "-D" "toolbindir=D:/a/dynamorio/dynamorio/build_debug-internal-64/bin64" "-D" "cmd=D:/a/dynamorio/dynamorio/build_debug-internal-64/bin64/run_in_bg@-out@D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/client.attach_test-out@D:/a/dynamorio/dynamorio/build_debug-internal-64/bin64/drrun.exe@-pidfile@D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/client.attach_test-pid@-no_inject@--@D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/bin/win32.infloop.exe" "-D" "out=D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/client.attach_test-out" "-D" "nudge=<attach>@-quiet@-debug@-use_dll@D:/a/dynamorio/dynamorio/build_debug-internal-64/lib64/debug/dynamorio.dll@-exit0@-stderr_mask@0xC@-msgbox_mask@0@-dumpcore_mask@0x7d@-staged@-code_api@-c@D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/bin/client.attach_test.dll.dll" "-D" "clear=" "-D" "pidfile=D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/client.attach_test-pid" "-P" "D:/a/dynamorio/dynamorio/suite/tests/runall.cmake"
2021-09-10T23:43:19.4680091Z 108: Test timeout computed to be: 600
2021-09-10T23:46:39.6901312Z 108: CMake Error at D:/a/dynamorio/dynamorio/suite/tests/runall.cmake:268 (message):
2021-09-10T23:46:39.6902351Z 108:   Timed out waiting for "done"
```

There's just not enough info there to know whether the attach succeeded and DR then hung, or what.  I would temporarily add more verbosity: `-stderr_mask 15` to get output on DR initializing; status messages from drrun on success of the attach; etc.  (If doing a lot of diagnostic runs maybe temporarily delete all the other `.github/workflows/ci-*.yml` files to avoid running everything else every time).  If it still passes locally but not on GA CI and verbosity is not enough you would have to resort to using tmate to log in and debug that way: https://dynamorio.org/page_test_suite.html#autotoc_md261

Reply:
Fixed it :)
The issue was that `win32.infloop.exe` was blocking indefinitely, and the attach failed.
I changed the target to the new (and simpler) `win32.attachee.exe` and it works.

Reply:
> The issue was that `win32.infloop.exe` was blocking indefinitely, and the attach failed.

But we want to be able to attach no matter what the target is doing.  Is this a limitation of DebugActiveProcess?  We need to document what the precise limitations are and we should file an issue on them with ideas on how to solve them.  Not being able to attach to a tiny process seems like a significant limitation.

Reply:
Attach failed when the main thread (to which we are injecting) was blocking.

I created a new suspended thread, and injected into it, and the attach is working for `win32.infloop.exe` as well.
I created the thread so that it would sleep indefinitely, because the assert in core/win32/syscall.c:2012 fails if the injected thread exists before the process does.

One thing I don't understand (and the reason I still cannot test with `infloop`) is closing the message box. How is the message box closed in all the tests? Who is responsible for it?

Reply:
> One thing I don't understand (and the reason I still cannot test with infloop) is closing the message box. How is the message box closed in the nudge test? Who is responsible for it?

It is closed by its window title with the closewnd utility here: https://github.com/DynamoRIO/dynamorio/blob/master/suite/tests/runall.cmake#L127